File: 1.BinaryToDecimal/PYTHON/main.py
from BinaryToDecimal import BinaryToDecimal
import json
import os

class Util:
    @staticmethod
    def read_json(file_path):
        with open(file_path, 'r') as file:
            return json.load(file)

    @staticmethod
    def write_json(file_path, data):
        with open(file_path, 'w') as file:
            json.dump(data, file, indent=4)

class BinaryToDecimalTest:
    def __init__(self):
        self.util = Util()
        self.config = self.util.read_json('conf/config.json')
        self.testcases = self.util.read_json(os.path.join('resources', 'testcases.json'))
        self.converter = BinaryToDecimal()

    def test(self, input, expected_output):
        result = self.converter.binary_to_decimal(int(input))
        return result == expected_output

def main():
    tester = BinaryToDecimalTest()
    testcasesJson = tester.testcases

    for testcase in testcasesJson:
        res = tester.test(testcase['input'], testcase['output'])
        testcase["result"] = "fail" if not res else "pass"

    # Update the result in python_result.json
    tester.util.write_json(tester.config.get('PythonResultFile'), testcasesJson)

    if "fail" in [testcase["result"] for testcase in testcasesJson]:
        raise Exception("Testcases failed")
    print("All testcases passed")

if __name__ == "__main__":
    main()

File: 1.BinaryToDecimal/CPP/src/main.cpp
#include "../include/json.hpp"
#include "../include/BinaryToDecimal.hpp"
#include "../include/BinaryToDecimalTest.hpp"
#include <fstream>
#include <iostream>
#include <vector>
#include <string>
#include <iomanip>
#include <unistd.h> 

using json = nlohmann::json;

class Config {
public:
    std::string testcasesFile;
    std::string CppResultFile;
};

class TestCase {
public:
    std::string input;
    int output;
    bool result;

    void to_json(json& j) {
        j = json{{"input", input}, {"output", output}, {"result", result}};
    }
};

json serialize(std::vector<TestCase> testcases) {
    json outputJson = json::array();
    for (TestCase testcase : testcases) {
        json j;
        testcase.to_json(j);
        outputJson.push_back(j);
    }
    return outputJson;
}

class Util {
public:
    std::vector<TestCase> testcases;
    Config config;

    Util() {
        config = readConfig();
        testcases = readTestcases();
    }

    Config readConfig() {
        std::ifstream configFile;
        configFile.open("../../conf/config.json");
        json configJson;
        configFile >> configJson;
        configFile.close();
        Config config;
        config.testcasesFile = configJson["testcasesFile"];
        config.CppResultFile = configJson["CppResultFile"];
        return config;
    }

    std::vector<TestCase> readTestcases() {
        std::ifstream testcaseFile;
        testcaseFile.open("../../" + config.testcasesFile);
        json j;
        testcaseFile >> j;
        testcaseFile.close();
        std::vector<TestCase> testcases;
        for (int i = 0; i < j.size(); i++) {
            TestCase testcase;
            testcase.input = j[i]["input"].get<std::string>();
            testcase.output = j[i]["output"].get<int>();
            testcases.push_back(testcase);
        }
        return testcases;
    }

    void writeResult() {
        std::ofstream resultFile;
        resultFile.open("../../" + config.CppResultFile);
        json j = serialize(testcases);
        resultFile << j.dump(2);
        resultFile.close();
    }
};

int main() {
    Util util;
    BinaryToDecimalTester tester;
    std::vector<TestCase>& testcases = util.testcases;
    bool failed = false;

    // Create a vector of pairs to hold the test cases
    std::vector<std::pair<std::string, int>> testCases;

    for (const TestCase& testcase : testcases) {
        // Add each test case to the vector of pairs
        testCases.emplace_back(testcase.input, testcase.output);
    }

    // Call the test method with the vector of test cases
    bool res = tester.test(testCases);

    failed = !res;

    for (std::size_t i = 0; i < testcases.size(); ++i) {
        // Update the result in the original TestCase object
        testcases[i].result = testCases[i].second == 1; // You can use a suitable condition here
    }

    util.writeResult();

    if (failed) {
        throw std::runtime_error("Some test cases failed");
    }

    std::cout << "All test cases passed" << std::endl;

    return 0;
}

